### PR TITLE
marks airdrops claimed or not

### PR DESCRIPTION
**Pull request by dimyG**
Backend part for #2000

## Checklist

- [ ] The PR modified the frontend, and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.

### CODE REVIEW

File: rotkehlchen/chain/ethereum/airdrops.py [hunk 1]
```diff
@@ -282,9 +289,36 @@ def get_poap_airdrop_data(name: str, data_dir: Path) -> dict[str, Any]:
...
+    )
+    with database.conn.read_ctx() as cursor:
+        events = DBHistoryEvents(database).get_history_events(
+            cursor=cursor,
+            filter_query=db_filter,
+            has_premium=True,  # no need to limit filtered events here
+        )
+        for event in events:
+            if event.balance.amount == amount:
+                return True
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/airdrops.py:314]: to avoid this loop you could add a filter on the amounts to the superclass of `EvmEventFilterQuery` and then use `get_history_events_count` so you could just check that 

```
return DBHistoryEvents(database).get_history_events_count(cursor, db_filter)[0] != 0
```

File: rotkehlchen/tests/unit/test_ethereum_airdrops.py [hunk 2]
```diff
@@ -28,10 +36,37 @@ def _fixture_airdrop_list(airdrop_list):
 
 @pytest.mark.parametrize('number_of_eth_accounts', [2])
 @pytest.mark.parametrize('use_clean_caching_directory', [True])
-def test_check_airdrops(ethereum_accounts, data_dir):
+def test_check_airdrops(ethereum_accounts, data_dir, database):
+    claim_event = EvmEvent(
+        tx_hash=deserialize_evm_tx_hash('0x5edc23d5a05e347afc60e64a4d5831ed2551985c21dceb85d267926ca2e2c13e'),  # A random tx_hash  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_ethereum_airdrops.py:41]: use

```suggestion
        tx_hash=make_evm_tx_hash()
```

File: rotkehlchen/chain/ethereum/airdrops.py [hunk 3]
```diff
@@ -282,9 +289,36 @@ def get_poap_airdrop_data(name: str, data_dir: Path) -> dict[str, Any]:
     return data_dict
 
 
+def is_airdrop_claimed(
+        address: ChecksumEvmAddress,
+        amount: FVal,
+        asset: Asset,
+        database: DBHandler,
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/airdrops.py:296]: since this logic is executed inside a nested loop consider creating `events_db = DBHistoryEvents(database)` outside in `def check_airdrops` and use it here. for the cursor you can do `events_db.db.conn.read_ctx()` since is an attribute. Doing this you avoid creating an object `NUMBER_OF_AIDROPS * NUMBER OF ADDRESSES` times

**Review by yabirgb — COMMENTED** [review]
This is missing a chagelog entry

File: rotkehlchen/db/filtering.py [hunk 4]
```diff
@@ -989,6 +990,14 @@ def make(
...
                 operator='NOT IN',
             ))
+        if amounts is not None:
+            str_amounts = [str(amount) for amount in amounts]
+            filters.append(DBMultiStringFilter(
+                and_op=True,
+                column='amount',
+                values=str_amounts,
+                operator='IN',
+            ))
```
Comment (yabirgb) [inline rotkehlchen/db/filtering.py:1000]: Something minor but you can just do

```suggestion
            filters.append(DBMultiStringFilter(
                and_op=True,
                column='amount',
                values=[str(amount) for amount in amounts],
                operator='IN',
            ))
```

**Review by yabirgb — COMMENTED** [review]
just something minor

File: rotkehlchen/api/rest.py [hunk 5]
```diff
@@ -2224,6 +2224,7 @@ def get_ethereum_airdrops(self) -> dict[str, Any]:
             data = check_airdrops(
                 addresses=self.rotkehlchen.chains_aggregator.accounts.eth,
                 data_dir=self.rotkehlchen.data_dir,
+                database=self.rotkehlchen.data.db,
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:2227]: Can't check as still mobile, but doesn't the database also contain the data directory or not? If it does perhaps you can remove it from above as an argument.

**Review by LefterisJP — COMMENTED** [review]
Left some comments

File: rotkehlchen/chain/ethereum/airdrops.py [hunk 6]
```diff
@@ -282,16 +290,52 @@ def get_poap_airdrop_data(name: str, data_dir: Path) -> dict[str, Any]:
...
+        addresses=addresses,
+        assets=tuple(assets),
+        amounts=amounts,
+    )
+    events_db = DBHistoryEvents(database)
+    with database.conn.read_ctx() as cursor:
+        events = events_db.get_history_events(
+            cursor=cursor,
+            filter_query=db_filter,
+            has_premium=True,
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/airdrops.py:311]:
```suggestion
            has_premium=True,  # needed for completeness sake anyway here
```

File: rotkehlchen/db/filtering.py [hunk 7]
```diff
@@ -568,6 +569,30 @@ def prepare(self) -> tuple[list[str], list[T]]:
...
+        parameters = []
+
+        for val in self.values:
+            value = val[0]
+            tolerance = val[1] * FVal(0.5)
+            conditions.append(f'CAST({self.column} AS REAL) BETWEEN ? AND ?')
+            parameters.extend([str(value - tolerance), str(value + tolerance)])
+
+        return (
+            [' OR '.join(conditions)],
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:591]: Why OR? Shouldn't it be AND? Or at least make this configurable? It already is an argument and you are giving it. The `and_op=True` one.

File: rotkehlchen/types.py [hunk 8]
```diff
@@ -881,6 +881,11 @@ def deserialize_from_db(cls, entry: tuple[int, str, str, str, int, int]) -> 'Use
         )
 
 
+class FValWithTolerance(NamedTuple):
```
Comment (LefterisJP) [inline rotkehlchen/types.py:884]: A small description of what this does in a docstring. Oneliner

File: rotkehlchen/types.py [hunk 9]
```diff
@@ -881,6 +881,11 @@ def deserialize_from_db(cls, entry: tuple[int, str, str, str, int, int]) -> 'Use
...
+    value: FVal
+    tolerance: FVal = FVal('0')
```
Comment (LefterisJP) [inline rotkehlchen/types.py:886]:
```suggestion
    tolerance: FVal = ZERO
```

File: rotkehlchen/db/filtering.py [hunk 10]
```diff
@@ -568,6 +569,30 @@ def prepare(self) -> tuple[list[str], list[T]]:
...
+    column: str
+    values: list[FValWithTolerance]
+
+    def prepare(self) -> tuple[list[str], list[str]]:
+        conditions = []
+        parameters = []
+
+        for val in self.values:
+            value = val[0]
+            tolerance = val[1] * FVal(0.5)
```
[the remark this answers is not in the document]
Comment (dimyG) [reply rotkehlchen/db/filtering.py:586]: Because the range is created by adding half the tolerance to the given value and removing half the tolerance from it so that the total tolerance is the given tolerance number. 
`parameters.extend([str(value - tolerance), str(value + tolerance)])`

File: rotkehlchen/db/filtering.py (hunk 7 above)
Comment (dimyG) [reply rotkehlchen/db/filtering.py:591]: It needs to be OR in this case because we query for a list of amounts. here is an example: 
image
That being said, I updated the filter to make the operation (AND or OR) configurable using the and_op parameter.

File: rotkehlchen/types.py (hunk 9 above)
Comment (dimyG) [reply rotkehlchen/types.py:886]: Moved some things around, to solve circular dependency issues for importing ZERO here.

File: rotkehlchen/types.py [hunk 11]
```diff
@@ -199,6 +200,8 @@ def deserialize_evm_tx_hash(val: Union[Web3HexBytes, bytearray, bytes, str]) ->
 T_TradeID = str
 TradeID = NewType('TradeID', T_TradeID)
 
+ZERO_PRICE = Price(ZERO)
```
Comment (LefterisJP) [inline rotkehlchen/types.py:203]: Why did you move this here? This is a constant, not a type. It should be inside the constants where it was before.

File: rotkehlchen/constants/misc.py [hunk 12]
```diff
@@ -20,8 +16,6 @@
 KRAKEN_BASE_URL = 'https://api.kraken.com'
 KRAKEN_API_VERSION = '0'
 
-ASSET_TYPES_EXCLUDED_FOR_USERS = {AssetType.NFT}
```
Comment (LefterisJP) [inline rotkehlchen/constants/misc.py:23]: This is fine moving. But not the zero price.

File: rotkehlchen/db/filtering.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/db/filtering.py:586]: I see. This makes sense. Can you maybe add a tiny comment next to it explaining this. In case others reading the code have the same question?

File: rotkehlchen/db/filtering.py (hunk 7 above)
Comment (LefterisJP) [reply rotkehlchen/db/filtering.py]: In that case there is a problem. This query should match each individual claim event. it does not.
Could also potentially match mistaken events if another event has same asset or any other combination of things here.


 I don't think our filters can work here. Cause you are trying to apply a multifilter. I really think raw sqlite is better here as I had written on Friday.

File: rotkehlchen/tests/unit/test_ethereum_airdrops.py [hunk 13]
```diff
@@ -28,10 +37,46 @@ def _fixture_airdrop_list(airdrop_list):
 
 @pytest.mark.parametrize('number_of_eth_accounts', [2])
 @pytest.mark.parametrize('use_clean_caching_directory', [True])
-def test_check_airdrops(ethereum_accounts, data_dir):
+def test_check_airdrops(ethereum_accounts, database):
+    # create airdrop claim events to test the claimed attribute
+    tolerance_for_amount_check = FVal('0.1')
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_ethereum_airdrops.py:42]: Shouldn't tolerance be smaller? Or wait .. this is just to test the tolerance attribute right?

**Review by LefterisJP — COMMENTED** [review]
Some minor comments and questions

**Review by LefterisJP — APPROVED** [review]
LGTM. Thanks.
